### PR TITLE
fmhalib

### PR DESCRIPTION
**Pull request by yjk21**
Optimized mha kernels for BERT

### CODE REVIEW

**Comment by yjk21** [comment]
FYI @slayton58 @seryilmaz

**Comment by slayton58** [comment]
\cc  @ptrblck @csarofeen

File: apex/contrib/csrc/fmha/src/fmha/philox.h [hunk 1]
```diff
@@ -0,0 +1,96 @@
+/***************************************************************************************************
+ * This code was copied from APEX (https://github.com/NVIDIA/apex). 
+ *
+ * APEX is released under BSD-3 license.
+ *
+ **************************************************************************************************/
+#pragma once
+// Philox CUDA.
+
+class Philox {
```
Comment (csarofeen) [inline apex/contrib/csrc/fmha/src/fmha/philox.h:10]: do we need a copy of this? CC @mcarilli

Comment (mcarilli) [reply apex/contrib/csrc/fmha/src/fmha/philox.h:10]: Looks like it's modified to do a "7-round" philox. Pytorch and curand do a "10-round" philox. No idea what the mathematical implications are, but this doesn't do the same thing as pytorch and curand philox. If they want it to do the 7 round thing they need their own copy.

Comment (seryilmaz) [reply apex/contrib/csrc/fmha/src/fmha/philox.h:10]: This also returns float4, whereas the original code returns float.

Comment (yjk21) [reply apex/contrib/csrc/fmha/src/fmha/philox.h:10]: I forgot to remove it. It's actually coming from here: https://github.com/NVIDIA/apex/blob/master/apex/contrib/csrc/multihead_attn/philox.h

Comment (seryilmaz) [reply apex/contrib/csrc/fmha/src/fmha/philox.h:10]: Right, and that one is coming from pytorch philox that @mcarilli  was referring to above.

File: apex/contrib/csrc/fmha/fmha_api.cpp [hunk 2]
```diff
@@ -0,0 +1,281 @@
+/***************************************************************************************************
+ * Copyright (c) 2011-2021, NVIDIA CORPORATION.  All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without modification, are not permit-
+ * ted.
```
Comment (cliffwoolley) [inline apex/contrib/csrc/fmha/fmha_api.cpp:5]: This is not an appropriate license for an open-source project; please fix to match APEX -- but also please ensure that doing so is permissible based on the original source of the code being contributed.

(Same comment for all files.)

**Review by cliffwoolley — CHANGES_REQUESTED** [review]

**Comment by ngoyal2707** [comment]
@yjk21
Curious what speed up this gets compared to `fast_multihead_attention`? and curious if there are plans to add `1024` sequence length?

**Pull request merged**